lib/at_parser/at_parser.c

```diff
@@ -0,0 +1,532 @@
+static bool lookahead_crlf_and_more(const char *str)
+{
+	return str[0] == CR && str[1] == LF && str[2] != NULL_TERMINATOR;
+}
+
+/* Lookahead if the string matches CRLF or starts with CRLF and has more data. */
+static bool lookahead_crlf_or_more(const char *str)
+{
+	return lookahead_crlf(str) || lookahead_crlf_and_more(str);
+}
+
+static void at_token_set_empty(struct at_token *token, const char *start)
+{
+	token->start = start;
+	token->len = 0;
+	token->type = AT_TOKEN_TYPE_EMPTY;
+	token->variant = AT_TOKEN_VARIANT_NO_COMMA;
+}
+
+static int at_parser_check(struct at_parser *parser)
+{
+	if (!parser) {
+		return -EINVAL;
+	}
+
+	if (!parser->at || !parser->cursor || parser->init_sentinel != INIT_SENTINEL) {
+		return -EPERM;
+	}
+
+	return 0;
+}
+
+/* Retrieve one token from the AT parser.  */
+static int at_parser_tok(struct at_parser *parser, struct at_token *token)
+{
+	const char *remainder = NULL;
+
+	/* The lexer cannot match empty strings, so intercept the special case where the empty
+	 * subparameter is the one after the previously parsed token.
+	 * This case is detected in the previous call to this function.
+	 */
+	if (parser->is_next_empty) {
+		at_token_set_empty(token, parser->cursor);
+
+		/* Set the remainder to the current cursor because an empty token has length
+		 * zero.
+		 */
+		remainder = parser->cursor;
+
+		/* Reset the flag. */
+		parser->is_next_empty = false;
+
+		goto finalize;
+	}
+
+	/* Trim leading CRLF. */
+	if (parser->count == 0) {
+		if (lookahead_crlf_and_more(parser->cursor)) {
+			parser->cursor += 2;
+		}
+	}
+
+	/* Nothing left to tokenize. */
+	if (parser->cursor[0] == NULL_TERMINATOR ||
+	    lookahead_cr_lf(parser->cursor) ||
+	    lookahead_crlf(parser->cursor)) {
+		return -EIO;
+	}
+
+	/* Don't tokenize response. */
+	if (is_resp(parser->cursor)) {
+		return -EIO;
+	}
+
+	/* Stop if there is a new line to parse. */
+	if (lookahead_cr_lf_and_more(parser->cursor) ||
+	    lookahead_crlf_and_more(parser->cursor)) {
+		return -EAGAIN;
+	}
+
+	if (parser->count == 0) {
+		*token = at_match_cmd(parser->cursor, &remainder);
+	} else {
+		*token = at_match_subparam(parser->cursor, &remainder);
+	}
+
+	if (token->type == AT_TOKEN_TYPE_INVALID) {
+		/* If the token is neither a valid command nor a valid subparameter, check if it's a
+		 * non-quoted string.
+		 */
+		*token = at_match_str(parser->cursor, &remainder);
+		if (token->type == AT_TOKEN_TYPE_INVALID) {
+			return -EBADMSG;
+		}
+	}
+
+	if (is_subparam(token)) {
+		/* if the token is a subparameter but does not have a trailing comma then it is the
+		 * last subparameter of the line.
+		 * If there is more to parse, then the token must be followed by CRLF, CR, or LF,
+		 * otherwise the string is malformed.
+		 */
+		bool lookahead = lookahead_cr_lf_or_more(remainder) ||
+				 lookahead_crlf_or_more(remainder);
+
+		if (token->variant != AT_TOKEN_VARIANT_COMMA) {
+			if (remainder[0] != NULL_TERMINATOR && !lookahead) {
```

Comment:
Tried to condense this logic a bit since I found parts of it hard to read. Got this:
```
bool next_is_NULL_CR_or_LF = (remainder[0] == NULL_TERMINATOR) ||
			     lookahead_cr_lf_or_more(remainder) ||
			     lookahead_crlf_or_more(remainder);

if (token->variant == AT_TOKEN_VARIANT_COMMA && next_is_NULL_CR_or_LF) {
	parser->is_next_empty = true;
} else if (token->variant == AT_TOKEN_VARIANT_NO_COMMA && !next_is_NULL_CR_or_LF) {
	return -EBADMSG;
}
```
I left out the comments in the snippet.